core/vm/instructions.go

```diff
@@ -262,6 +261,12 @@ func opAddress(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]
 func opBalance(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]byte, error) {
 	slot := scope.Stack.peek()
 	address := common.Address(slot.Bytes20())
+	if interpreter.evm.chainRules.IsVerkle {
+		chargedGas, ok := interpreter.evm.Accesses.TouchBasicData(address[:], false, scope.Contract.UseGas)
+		if !ok || (chargedGas == 0 && !scope.Contract.UseGas(params.WarmStorageReadCostEIP2929)) {
```

Comment:
You could be wondering why some of the `Touch*(...)` APIs returned the charged gas if we were sending the callback?

This is the reason. We have this rule that if on many charges the actual charged gas was 0, then we charge warm. We ned a way to know if that `UseGas` was asked to charge 0 or not.

If was 0, we do a usual `scope.Contract.UseGas(Warm)`.